readme.md
## RESO Feed Faker
[![Latest Version](https://img.shields.io/github/release/sparkapi/feed-faker.svg?style=flat-square)](https://github.com/sparkapi/feed-faker/releases)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://img.shields.io/travis/sparkapi/feed-faker/master.svg?style=flat-square)](https://travis-ci.org/sparkapi/feed-faker)
[![Total Downloads](https://img.shields.io/packagist/dt/sparkapi/feed-faker.svg?style=flat-square)](https://packagist.org/packages/sparkapi/feed-faker)

FeedFaker allows you to generate a ton of fake RESO-compliant feed data objects quickly and easily, in a way that will keep the IDs at least semi-referentially linked for imports into databases.

## Install via Composer
`$ composer require sparkapi/feed-faker`

(or)
```json
{
    "require": {
        "sparkapi/feed-faker": "^0.8"
    }
}
```

## Example Usage

First define your settings for keys and the number and type of objects to create.

If you are using the unsplash photo provider you will have to pass in a few additional settings.
If you want to use a different image provider, pass in the class to that provider.

All other settings have defaults, so feel free to modify what you want and leave blank the rest.

```php
$settings = [
    // Only required for unsplash
    'unsplash_app_id' => 'your_unsplash_app_id', // unsplash specific
    'interior_collection' => collection_id, // unsplash specific
    'exterior_collection' => collection_id, // unsplash specific
    'portrait_collection' => collection_id, // unsplash specific
        
    // Optional photo provider, if not using unsplash (you can ignore the above, too).
    'photo_provider' => \FeedFaker\Helper\UnsplashPhotoProvider::class,

    // Basic settings, all with defaults.
    'mls_id' => 'my-awesome-mls',
    
    // the number of objects you want to create for each resource type.
    'office_count' => 50,
    'member_count' => 150,
    'property_count' => 500,
    'openhouse_count' => 1000,
    'history_count' => 1000,
    'media_count' => 5000,
];
```

You may provide a valid full path to a php settings file that returns an array, if you prefer. An example is in `settings.php.dist`.

Now that you're all setup, instantiate it and fire away:

```php
$settings = [
    'unsplash_app_id' => 'my-app-id-from-unsplash'
    'interior_collection' => 10
    'exterior_collection' => 1010
    'portrait_collection' => 101010
    
    // optional
    'mls_id' => 'best-mls-in-the-universe',
    'history_count' => 500
];

$feedfake = new FeedFaker\FakeFeed();
$data = $feedfake->go($settings);

$properties = $data['property'];
$offices = $data['office'];
$members = $data['member'];
$media = $data['media'];
$history = $data['history'];
$openhouses = $data['openhouse'];

foreach ($properties as $property) {
    echo $property->getListingKey() . PHP_EOL;

    var_dump($property->toArray());
}
```

Yep, that's it. You're welcome.
    
## Optional Logging
The Faker can also log information to any PSR-3 compatible logger. (Useful for Image provider errors, etc.)

You have two options to set one up: Give it **your own logger**, or provide a setup **callable** for MonoLog.

### Setup Callable for Monolog
By default, Faker uses Monolog to handle any logging, which uses handlers to know where and what details to log. 
If you provide no setup or logger, Faker will default to a `NullHandler` and save no log information.

You may provide any callable as the second parameter for `FakeFeed`. This callable accepts a monolog `Logger` instance and doesn't need to return anything.

```php
$logger_setup = function($monolog) {
    // Save all debug and above messages to a file
    $monolog->pushHandler(new StreamHandler('/path/to/logs.log', \Monolog\Logger::DEBUG));
    
    // Save more important messages to some other place
    $monolog->pushHandler(new SomeOtherHandler(\Monolog\Logger::WARNING));
};

$feedfake = new FeedFaker\FakeFeed([], $logger_setup);
$data = $feedfake->go()
```

### Your own logger
You may also skip monolog altogether and provide any instance of a PSR-3 compatible logger as the second argument.

```php
$feedfake = new FeedFaker\FakeFeed([], $my_psr_3_logger);
$data = $feedfake->go()
```

## Interoperability
Feed Faker is [PSR compliant](http://www.php-fig.org/) and [Container Interoperability](https://github.com/container-interop/container-interop) compliant. Any oversights, please let us know.

## Testing
We try for at least 80% test coverage.
``` bash
$ phpunit
```

Note tht the `UnsplashPhotoProviderTest`s are skipped. These require a valid credential. You may extend and run these tests with your own credentials. A better (environment-based) system is on the way.

## Contributing
Contributions are welcome and will be fully credited. Please see [CONTRIBUTING](CONTRIBUTING.md) for details.

Open an issue to join in!

## License
The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


changelog.md
# Changelog
All Notable changes to `Feed Faker` will be documented in this file

## Next

## v0.8 - 2017-4-27
This was the first public release version. Introduces Faking, Providers, Settings, Logging, and Tests.


src/Config/Settings.php
<?php

namespace FeedFaker\Config;

use FeedFaker\Providers\UnsplashPhotoProvider;
use FeedFaker\Providers\NullPhotoProvider;

/**
 * Class Settings
 * @package FeedFaker\Config
 */
class Settings {

    /** @var array */
    protected static $defaults = [
        'mls_id' => 'awesome-mls',
        'office_count' => 50,
        'member_count' => 150,
        'property_count' => 500,
        'openhouse_count' => 1000,
        'history_count' => 1000,
        'media_count' => 5000,
        'photo_provider' => NullPhotoProvider::class,
    ];

    /**
     * Take the given settings and return valid settings with defaults
     *
     * @param $config
     * @return array
     * @throws \Exception
     */
    public static function load($config)
    {
        // Validate settings
        if (!is_array($config) && !is_string($config)) {
            throw new \Exception("FakeFeed requires a valid array of settings or the path to a php settings file.");
        }

        // Load a directory
        if (is_string($config)) {
            if  (!is_readable($config)) {
                throw new \Exception("Settings path {$config} is not readable or does not exist");
            }

            /** @var array $config */
            $config = require $config;
        }

        // Merge in defaults
        return static::validate(array_merge(static::getDefaults(), $config));
    }

    /**
     * Validate the loaded settings for required fields
     *
     * @param array $settings
     * @return array
     * @throws \Exception
     */
    public static function validate(array $settings)
    {
        if (
            // If we are using Unsplash, make sure the settings are valid
            // This may be better in the Unsplash provider some time in the future.
            $settings['photo_provider'] === UnsplashPhotoProvider::class
            && (
                (!array_key_exists('unsplash_app_id', $settings) || !is_string($settings['unsplash_app_id']))
                || (!array_key_exists('interior_collection', $settings) || !is_int($settings['interior_collection']))
                || (!array_key_exists('exterior_collection', $settings) || !is_int($settings['exterior_collection']))
                || (!array_key_exists('portrait_collection', $settings) || !is_int($settings['portrait_collection']))
            )
        ){
            throw new \Exception("Given settings are missing required unsplash fields. Please see readme.md");
        }

        // If it passed, we just hand back the settings
        return $settings;
    }

    /**
     * @return array
     */
    public static function getDefaults()
    {
        return static::$defaults;
    }
}
